compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs

```diff
@@ -2001,6 +1986,75 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
         existing_bound_vars.extend(bound_vars);
         self.record_late_bound_vars(item_segment.hir_id, existing_bound_vars_saved);
     }
+
+    /// Walk the generics of the item for a trait-ref whose self type
+    /// corresponds to the expected res.
+    fn for_each_in_scope_predicate(
+        &self,
+        expected_res: Res,
+    ) -> impl Iterator<Item = &'tcx hir::PolyTraitRef<'tcx>> + use<'tcx, '_> {
+        std::iter::from_coroutine(
+            #[coroutine]
+            move || {
+                let mut next_scope = Some(self.scope);
+                while let Some(current_scope) = next_scope {
+                    next_scope = None;
+                    let hir_id = match *current_scope {
+                        Scope::Binder { s, hir_id, .. } => {
+                            next_scope = Some(s);
+                            hir_id
+                        }
+                        Scope::Body { s, .. }
+                        | Scope::ObjectLifetimeDefault { s, .. }
+                        | Scope::Supertrait { s, .. }
+                        | Scope::TraitRefBoundary { s }
+                        | Scope::LateBoundary { s, .. } => {
+                            next_scope = Some(s);
+                            continue;
+                        }
+                        Scope::Root { opt_parent_item } => {
+                            if let Some(parent_id) = opt_parent_item {
+                                self.tcx.local_def_id_to_hir_id(parent_id)
+                            } else {
+                                continue;
```

Comment:
just to make sure I'm following the logic correctly ...

On this branch, `next_scope` is still `None` (right?), and therefore this `continue` will end up sending a `None` up into the `while let Some(current_scope)` condition, which means that this `continue` effectively ends up acting like ... a `break`?

(But then ... why not just write `break` here?)

Reply:
In other methods, we prefer `scope` not being an option, and breaking when getting to `Root`. Do you mind making this a bit more uniform?